### ISSUE TITLE
Cannot remove array: not found in ArrayCache

### ISSUE THREAD

**Issue by tarilabs** [deeplearning4j/deeplearning4j#9044]
#### Issue Description

I am running a TF via `org.nd4j.autodiff.samediff.SameDiff` and I am interested only in a single output.

If I run as:
```java
Map<String, INDArray> out = sd.batchOutput().input("signal", tracings).outputAll().output();
```
it works and I can find the single output I'm particularly interested, in the resulting map.

If I run instead as:
```java
Map<String, INDArray> out = sd.batchOutput().input("signal", tracings).output("dense_8/Sigmoid").output();
```
I get the following stacktrace:
```
Exception in thread "main" java.lang.IllegalStateException: Cannot remove array: not found in ArrayCache
        at org.nd4j.common.base.Preconditions.throwStateEx(Preconditions.java:641)
        at org.nd4j.common.base.Preconditions.checkState(Preconditions.java:268)
        at org.nd4j.autodiff.samediff.internal.memory.ArrayCacheMemoryMgr$ArrayStore.removeObject(ArrayCacheMemoryMgr.java:261)
        at org.nd4j.autodiff.samediff.internal.memory.ArrayCacheMemoryMgr$ArrayStore.access$100(ArrayCacheMemoryMgr.java:185)
        at org.nd4j.autodiff.samediff.internal.memory.ArrayCacheMemoryMgr.release(ArrayCacheMemoryMgr.java:146)
        at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:345)
        at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:68)
        at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:391)
        at org.nd4j.autodiff.samediff.SameDiff.directExecHelper(SameDiff.java:2579)
        at org.nd4j.autodiff.samediff.SameDiff.batchOutputHelper(SameDiff.java:2547)
        at org.nd4j.autodiff.samediff.SameDiff.output(SameDiff.java:2522)
        at org.nd4j.autodiff.samediff.config.BatchOutputConfig.output(BatchOutputConfig.java:143)
        at asd.App3.main(App3.java:24)
```

#### Version Information

* Deeplearning4j version `1.0.0-beta7`
* Platform information (OS, etc) `macOS 10.14.6`

#### Additional Information

pom versions:
```xml
    <dependency>
      <groupId>org.deeplearning4j</groupId>
      <artifactId>deeplearning4j-modelimport</artifactId>
      <version>1.0.0-beta7</version>
    </dependency>
    <dependency>
      <groupId>org.nd4j</groupId>
      <artifactId>nd4j-native</artifactId>
      <version>1.0.0-beta7</version>
    </dependency>
```

This is a follow-up problem to what originally described here: https://github.com/eclipse/deeplearning4j/issues/9043#issuecomment-660458324

here is the Java snippet I'm using:
```java
public static void main(String[] args) throws Exception {
        SameDiff sd = SameDiff.importFrozenTF(new File("../automatic-ecg-diagnosis/freeze_session.pb"));
        System.out.println(sd.summary());
        System.out.println(sd.inputs());
        Hdf5Archive hdf5Archive = new Hdf5Archive("../automatic-ecg-diagnosis/data/ecg_tracings.hdf5");
        System.out.println(hdf5Archive.getDataSets());
        INDArray tracings = hdf5Archive.readDataSet("tracings");
        // INDArray asd = tracings.get(NDArrayIndex.point(18), NDArrayIndex.all(), NDArrayIndex.all());
        // System.out.println(Arrays.toString(asd.shape()));
        Map<String, INDArray> out = sd.batchOutput().input("signal", tracings).output("dense_8/Sigmoid").output();
        for (Entry<String, INDArray> kv : out.entrySet()) {
            if (!kv.getKey().startsWith("dense")) {
                continue;
            }
            System.out.println(kv.getKey());
            System.out.println(kv.getValue());
        }
        System.out.println("dense_8/Sigmoid");
        System.out.println(out.get("dense_8/Sigmoid").toStringFull());
    }
```

To run this, will need `freeze_session.pb` and `ecg_tracings.hdf5` locally.
I summarize below how I obtained these two files.

##### Generate locally File("../automatic-ecg-diagnosis/freeze_session.pb")

As originally described here: https://github.com/eclipse/deeplearning4j/issues/9043#issuecomment-660458324

In short, the original model defined here: https://github.com/antonior92/automatic-ecg-diagnosis/blob/cb3394da2af36f1199b03736da87478c6cc31dfc/model.py#L117-L137
and a pre-trained model can be directly downloaded as explained here: https://github.com/antonior92/automatic-ecg-diagnosis#scripts
from this page: https://doi.org/10.5281/zenodo.3625017
Scrolling to the bottom you will find option to download `model.zip`
will give the `model.hdf5`.

To generate the `.pb` file follow the original python snippet as described here: https://github.com/eclipse/deeplearning4j/issues/9043#issuecomment-660458324
in short:
```python
import numpy as np
import tensorflow as tf
import h5py

tf.keras.backend.set_learning_phase(0)
model = tf.keras.models.load_model('./dnn_predicts/model.hdf5')

def freeze_session(session, keep_var_names=None, output_names=None, clear_devices=True):
    """
    Freezes the state of a session into a pruned computation graph.

    Creates a new computation graph where variable nodes are replaced by
    constants taking their current value in the session. The new graph will be
    pruned so subgraphs that are not necessary to compute the requested
    outputs are removed.
    @param session The TensorFlow session to be frozen.
    @param keep_var_names A list of variable names that should not be frozen,
                          or None to freeze all the variables in the graph.
    @param output_names Names of the relevant graph outputs.
    @param clear_devices Remove the device directives from the graph for better portability.
    @return The frozen graph definition.
    """
    graph = session.graph
    with graph.as_default():
        freeze_var_names = list(set(v.op.name for v in tf.global_variables()).difference(keep_var_names or []))
        output_names = output_names or []
        output_names += [v.op.name for v in tf.global_variables()]
        input_graph_def = graph.as_graph_def()
        if clear_devices:
            for node in input_graph_def.node:
                node.device = ''
        frozen_graph = tf.graph_util.convert_variables_to_constants(
            session, input_graph_def, output_names, freeze_var_names)
        return frozen_graph

frozen_graph = freeze_session(tf.keras.backend.get_session(), output_names=[out.op.name for out in model.outputs])
tf.train.write_graph(frozen_graph, './', 'freeze_session.pbtxt', as_text=True)
tf.train.write_graph(frozen_graph, './', 'freeze_session.pb', as_text=False)
```

##### Download locally Hdf5Archive("../automatic-ecg-diagnosis/data/ecg_tracings.hdf5")

As originally described here: https://github.com/eclipse/deeplearning4j/issues/9043#issuecomment-660610459

In short, as described in the original source https://github.com/antonior92/automatic-ecg-diagnosis#scripts from this archive: https://doi.org/10.5281/zenodo.3625006
At the bottom of the page, find option to download `data.zip`
Inside it can be found `data/ecg_tracings.hdf5`

##### Reproducing stacktrace

With all the needed data files, the above java snippet can be run and will reproduce the stacktrace

##### Workaround, and other notes

If I change the line as explained in the introduction to:
```java
Map<String, INDArray> out = sd.batchOutput().input("signal", tracings).outputAll().output();
```

Now it will run completely and I will get the expected output

```
   (...snip...)
[signal]
[tracings]
dense_8/BiasAdd/ReadVariableOp
[   -0.4971,    0.4819,   -0.3434,   -0.1355,   -0.5597,   -0.0892]
dense_8/kernel
[[   -0.0596,   -0.0073,   -0.0206,    0.0739,   -0.0568,   -0.0172], 
 [   -0.8135,   -0.3057,   -0.3090,   -0.4890,   -0.4670,   -0.4100], 
 [    0.9735,    0.0080,    1.2367,    0.6044,   -7.9368,   -1.9769], 
  ..., 
 [   -0.0404,    0.0125,    0.0774,    0.0029,    0.0137,   -0.0254], 
 [   -0.0405,   -0.3509,    0.0565,   -0.0180,   -0.3206,   -0.3919], 
 [    0.1146,   -0.1248,   -0.1212,   -0.0429,   -0.0284,   -0.0501]]
dense_8/bias
[   -0.4971,    0.4819,   -0.3434,   -0.1355,   -0.5597,   -0.0892]
dense_8/MatMul
[[  -12.9647,  -16.5314,  -14.8063,  -14.4702,  -13.3086,  -18.7756], 
 [   -3.0176,   -6.6912,   -0.4206,  -10.3379,   -2.4201,   -7.9561], 
 [   -7.5775,  -10.9163,  -12.0430,  -10.6987,   -4.1034,  -10.5101], 
  ..., 
 [  -13.8163,  -17.5200,  -17.3913,  -18.1967,  -13.2384,  -17.0485], 
 [  -16.5519,  -20.3274,  -20.5602,  -18.3076,  -15.9436,  -19.2597], 
 [   -1.0652,   -6.4884,   -9.0177,   -9.8437,   -5.6270,   -9.4389]]
dense_8/BiasAdd
[[  -13.4618,  -16.0495,  -15.1497,  -14.6057,  -13.8683,  -18.8649], 
 [   -3.5147,   -6.2093,   -0.7640,  -10.4734,   -2.9799,   -8.0453], 
 [   -8.0746,  -10.4344,  -12.3863,  -10.8342,   -4.6631,  -10.5993], 
  ..., 
 [  -14.3134,  -17.0381,  -17.7347,  -18.3322,  -13.7981,  -17.1377], 
 [  -17.0490,  -19.8455,  -20.9035,  -18.4431,  -16.5033,  -19.3490], 
 [   -1.5623,   -6.0065,   -9.3611,   -9.9792,   -6.1867,   -9.5282]]
dense_8/Sigmoid
[[ 1.4243e-6,  1.071e-7, 2.6337e-7, 4.5377e-7, 9.4854e-7, 6.4135e-9], 
 [    0.0289,    0.0020,    0.3178, 2.8277e-5,    0.0483,    0.0003], 
 [    0.0003, 2.9403e-5, 4.1752e-6, 1.9713e-5,    0.0093, 2.4932e-5], 
  ..., 
 [ 6.0783e-7, 3.9853e-8, 1.9858e-8, 1.0925e-8, 1.0176e-6, 3.6073e-8], 
 [  3.942e-8, 2.4055e-9,8.3505e-10, 9.7783e-9, 6.8028e-8, 3.9523e-9], 
 [    0.1733,    0.0025, 8.5999e-5, 4.6351e-5,    0.0021, 7.2769e-5]]
dense_8/MatMul/ReadVariableOp
[[   -0.0596,   -0.0073,   -0.0206,    0.0739,   -0.0568,   -0.0172], 
 [   -0.8135,   -0.3057,   -0.3090,   -0.4890,   -0.4670,   -0.4100], 
 [    0.9735,    0.0080,    1.2367,    0.6044,   -7.9368,   -1.9769], 
  ..., 
 [   -0.0404,    0.0125,    0.0774,    0.0029,    0.0137,   -0.0254], 
 [   -0.0405,   -0.3509,    0.0565,   -0.0180,   -0.3206,   -0.3919], 
 [    0.1146,   -0.1248,   -0.1212,   -0.0429,   -0.0284,   -0.0501]]
   (...snip...)
```

So not sure why indicating specifically only the output I'm interested in with `.output("dense_8/Sigmoid")` is causing the problem?

#### Contributing

The model I'm using is quite big so it makes difficult for me to run an A/B debug session to try help in understanding what's going on, but if you can put me on the right track to have a more sensible reproducer or where you believe the problem lies, I could try to be of support. Or if this is enough for you to quick fix, I'm glad if it helped! :)

Thanks!

**Comment by agibsonccc** [comment]
@tarilabs can you try a simpler workflow like this example?
https://github.com/eclipse/deeplearning4j-examples/blob/master/samediff-examples/src/main/java/org/nd4j/examples/samediff/quickstart/basics/Ex2_LinearRegression.java#L77

Generally when you want outputs, you specify which outputs you want to use. I suspect it's related to your usage of batchOutput. Generally, you shouldn't need that. Outputs as one ndarray should be pretty straightforward.

**Comment by agibsonccc** [comment]
@tarilabs any updates here?

**Comment by tarilabs** [comment]
sorry for the delay, this is more of a "weekend" interest activity for me, will try it out asap

**Comment by Scoobadood** [comment]
I saw the same issue when forward inferencing on a CNN built using only ND4J classes.

Tracing the code led me to this code and in particular line 255:

https://github.com/eclipse/deeplearning4j/blob/6b736eae3333a94c1ef22ae5e53b1139434bb4b5/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/memory/ArrayCacheMemoryMgr.java#L252-L263

```
 Preconditions.checkState(idx > 0, "Cannot remove array from ArrayStore: no array with this length exists in the cache");
```

In my case, an array of the appropriate size was present in `lengths` at index 0 but was explicitly failed by this test. FWIW, forcing that test to pass in the debugger fixed the problem superficially but I did not run regression testing of the whole suite so it may not be a long term solution.  Any thoughts or comments?  Is this test intentionally excluding the first element of the array?

**Comment by agibsonccc** [comment]
@Scoobadood Can you come up with a self contained network that contains this for a test case? If so, we can get a proper test case and run the regression tests relative to that.

**Comment by Scoobadood** [comment]
I can TAL. A challenge is forcing a memory shortage such that the cache is required to remove or recycle things (I don't have a deep understanding of how the cache works).  My network is not exactly simple and this particular issue only arises when my input image is "large enough". YMMV.  A unit test may be a better approach here.

**Comment by agibsonccc** [comment]
@Scoobadood yeah sure, anything that can reproduce this and would be considered a valid case will help get this fixed. It will also help force us to think about what is/isn't a valid case here. Thanks!

**Comment by partarstu** [comment]
@agibsonccc , I've got the same issue while working on my own `SameDiff `model. In my case it showed up while calling `SameDiff.grad("multi_head_dot_product_attention").eval()` using the same dataset which has been used for training as an input. My model is a sort of Transformer's encoder (similar to BERT's one) and the main part of it is self-attention (`MultiHeadDotProductAttention`). The trigger of the error was increasing the sequence length. Having the batch size of 512 and sequence length of 64 with embedding size (hidden size) of 768 was working fine. But after increasing the sequence length to 128 I got the failure. Decreasing the batch size from 512 to 256 having the same sequence length helped in my case, but not for a long time - after a couple of training rounds I still got the same error. I know that such parameters are quite huge for unit testing (in my case I had to turn on periodic GC and set the auto GC window to 20 seconds in order to avoid OOE having 32GB of RAM), but hopefully this info could give you a hint to finding the root cause.
